### ISSUE TITLE
Urban Airship needs to register channel id to backend

### ISSUE THREAD

**Issue by TheAznShumai** [blockmason/lndr-mobile#97]
We need to be able to send the channel ID to the backend so that we can use that id to send push notifications to users.  An Endpoint needs to be made for this.

**Comment by aupiff** [comment]
Please give me some more info on this so I can make the necessary backend endpoint

**Comment by TheAznShumai** [comment]
Spec

Some User Update Action (update_device?)
`{ 
  channelD: 'string',
  platform: 'string'
}`

If there is another endpoint we can take this data on, we can do that too.

**Comment by aupiff** [comment]
I'm still not sure what you need me to store on the server. Should I create a db table consisting of two columns, `channelID` and `platform`? Should `channelID` be unique? Do you want to be able to query this information somehow?

**Comment by TheAznShumai** [comment]
So what happens is there we will need to store this information with the association to the user such that you can use those items to do api requests to urban airship and it'll send notifications accordingly.  A use case example is on confirmation, I want to send to the associate whom the user confirmed and you will need to send an api request with the channel id to urban airship so it can follow through with pushing a notification on their device, assuming they have it enabled and stored on the backend.

Here is documentation for a basic push notification
https://docs.urbanairship.com/api/ua/#push-object

Example
```
{
    "audience": {
        "android_channel": "31279004-103e-4ba8-b4bf-65eb3eb81859"
    },
    "notification": {
        "alert": "hello world"
    },
    "device_types": [
        "android"
    ]
}
```

**Comment by aupiff** [comment]
I need a few more details.

- tell what what information I store in the db and what you want in terms of an endpoint
   + I am assuming that you'll want to store only `channelID` and `platform`, that `channelID` will be 
unique, and that these two fields should be associated with a user address

- What events are we supporting push notifications for? I'm assuming we'll start with:
    + new pending credit record
    + rejected credit record 
    + confirmed credit record

- give me dummy versions of all the urban airship requests we'll need for the mobile app. I assume you'll determine what needs to go in place of the `"notification": { "alert": "hello world"  }"`

- tell me what endpoint I should be hitting on urban airship's servers? Does our app get assigned a custom path in their api?

**Comment by timgalebach** [comment]
For the first 2, your assumptions are correct. I need Johnny to answer the latter ones.

**Comment by TheAznShumai** [comment]
I noted an example above.  You should be hitting the push endpoint with 
```
{
    "audience": {
        "android_channel": "31279004-103e-4ba8-b4bf-65eb3eb81859"
    },
    "notification": {
        "alert": "hello world"
    },
    "device_types": [
        "android"
    ]
}
```

The documentation is located at https://docs.urbanairship.com/api/ua/#push-api

**Comment by TheAznShumai** [comment]
I don't understand what you mean by dummy version of the request.  If you mean the actual message, then I don't have that text.

You can use the https://github.com/madelgi/ua-haskell-library wrapper since the backend is written in haskell from what I hear.  You should be hitting the push endpoint https://docs.urbanairship.com/api/ua/#push-api

**Comment by TheAznShumai** [comment]
Actually the wrapper looks like it pretty old and unsupported, but there might be good setup code that you could use.

**Comment by TheAznShumai** [comment]
so if you're talking about passing an event data to trigger the mobile app to do something we could do something like

```
{
   "audience": "all",
   "notification": {
      "alert": "You got your emails!",
      "actions": {
          "app_defined": { "LNDR_ACTIONS" : ["REFRESH_ACTIVITY"] }
       }
   },
   "device_types" : "all"
}
```
As you can see the app_defined stuff can be customizable and is sent mobile when push notification are active.  RIght now I'm thinking we should incorporate these types of custom action similar to how redux consume their actions within the LNDR engine.  I was thinking about maybe having the ability to execute multiple ACTIONS in sequence.

**Comment by aupiff** [comment]
Why would we want to execute multiple actions in sequence? Isn't the only action we're interested in creating a notification on a user's phone of some pending credit event?

**Comment by aupiff** [comment]
Do you want to limit the value of `platform` to 'android' and 'ios" or do you really want it be of type string?

**Comment by aupiff** [comment]
What sort of behavior should we plan for if the user is signed into multiple devices? The way I've written the first implementation of the `/register_push` endpoint, the most recently signed-into device will be the one that receives notifications.

Should I make a user capable of registering multiple (channelID, platforms) pairs?

**Comment by aupiff** [comment]
Here are the autogenerated docs for the `/register_push` endpoint: https://github.com/blockmason/lndr/tree/feature-urban-airship/lndr-backend#post-register_pushuser

`POST /register_push/:user` where `user` is an ethereum address. The json payload must be of the following format: `{"channelID":"31279004-103e-4ba8-b4bf-65eb3eb81859","platform":"ios"}`

Let me know if that works.

**Comment by TheAznShumai** [comment]
* The reason to do multiple action is to think of a philosophy you want to adopt in controlling the front end.  Would we want to control individual parts of the application based off an event that occurred in the backend?  or do we want to have the front end control that by have a generalized wrapper of actions to do if there was something complex needed.  
- An example is like maybe tell the frontend to refresh a section and also pull up an upsell component or something to that nature such that the background is refreshing and you get a popup.
** I almost feel like it might be easier for the backend to control things on this note.

*  For platform field validations, I realize that our upcoming work is only for ios and android, but technically we can have other platforms.  It should be fine to leave as is.  The only reason we have the platform string is to use to as a way to create the urbanairship api requests.  So if there is a platform that unexpected in the backend, we should gracefully capture it and not send the urbanairship request.

* This should work in doing storage.  You will be making the requests to urbanairship after all.  If there is some additional complexity, we can always change this.

**Comment by TheAznShumai** [comment]
Added a new card to continue work on the front end under
https://github.com/blockmason/lndr-mobile/issues/114

We'll need to figure out if the actions array is good way to go about things, create a new issue, and close out this issue.

**Comment by timgalebach** [comment]
@aupiff just make sure that the channel id registration endpoint is idempotent, so we don't get an error from hitting it twice with the same user.

**Comment by aupiff** [comment]
@timgalebach I've made it so users can set new channel ids. same logic as nicknames.

**Comment by willbach** [comment]
Fixed with PR #135
